### PR TITLE
Framerate converter

### PR DESCRIPTION
**Pull request by a-rose**
# PR Type

- [x] Feature
- [ ] Bug fix
- [ ] Docs

## Description

Add framerate conversion capabilities using libavfilter.
The framerate is automatically converted if the encoder's fps setting doesn't match exactly the input framerate

### CODE REVIEW

File: cgo/ffmpeg/video.go [hunk 1]
```diff
@@ -171,6 +196,176 @@ func (self *VideoScaler) VideoScale(src *VideoFrame) (dst *VideoFrame, err error
 	return
 }
 
+type FramerateConverter struct {
```
Comment (JulienBreux) [inline cgo/ffmpeg/video.go]: Best practices to comment your code.
(even if the lib have no comments)

File: cgo/ffmpeg/video.go [hunk 2]
```diff
@@ -171,6 +196,176 @@ func (self *VideoScaler) VideoScale(src *VideoFrame) (dst *VideoFrame, err error
...
+	inHeight, OutHeight int
+	inFpsNum, OutFpsNum int
+	inFpsDen, OutFpsDen int
+	pts int
+	graph *C.struct_AVFilterGraph
+	inVideoFilter  *C.AVFilterContext // the first filter in the video chain
+	outVideoFilter *C.AVFilterContext // the last filter in the video chain
+	outputFrames []*C.AVFrame
+}
+
```
Comment (JulienBreux) [inline cgo/ffmpeg/video.go]: Best practices to comment your code.
(even if the lib have no comments)
...

**Review by JulienBreux — CHANGES_REQUESTED** [review]

**Comment by JulienBreux** [comment]
@a-rose up for comments

**Comment by a-rose** [comment]
> @a-rose up for comments

I'm almost done cleaning up :)

File: cgo/ffmpeg/video.go [hunk 3]
```diff
@@ -18,9 +26,15 @@ import (
 	"github.com/nareix/joy4/codec/h264parser"
 )
 
+type VideoFramerate struct {
```
Comment (JulienBreux) [inline cgo/ffmpeg/video.go]: Comments here

File: cgo/ffmpeg/video.go [hunk 4]
```diff
@@ -56,6 +70,10 @@ func (v VideoFrame) GetDataPtr() (y, cb, cr *[]uint8) {
 	return &v.Image.Y, &v.Image.Cb, &v.Image.Cr
 }
 
+func (v VideoFrame) GetFramerate() (num, den int) {
```
Comment (JulienBreux) [inline cgo/ffmpeg/video.go]: Comments here

File: cgo/ffmpeg/video.go [hunk 5]
```diff
@@ -80,6 +98,11 @@ func (v *VideoFrame) SetResolution(w, h int) {
 	v.Image.Rect = image.Rectangle{ image.Point{0,0}, image.Point{w, h}}
 }
 
+func (v *VideoFrame) SetFramerate(num, den int) {
```
Comment (JulienBreux) [inline cgo/ffmpeg/video.go]: Comments here

**Review by JulienBreux — CHANGES_REQUESTED** [review]
Sorry but I am not made to understand ☺️

**Review by JulienBreux — APPROVED** [review]

**Comment by cedricve** [comment]
@a-rose  great. I'm just wondering how to properly calculate the FrameRate if this is note set.

https://github.com/Seize/joy4/blob/master/codec/h264parser/parser.go#L650

**Comment by a-rose** [comment]
> @a-rose great. I'm just wondering how to properly calculate the FrameRate if this is note set.
> 
> https://github.com/Seize/joy4/blob/master/codec/h264parser/parser.go#L650

If the timing information is not present, you're probably dealing with variable FPS, as is often the case with mobile video streaming.

**Pull request merged**